### PR TITLE
Make ToolCache thread-safe

### PR DESCRIPTION
**Pull request by ras0219-msft**

### CODE REVIEW

File: include/vcpkg/base/delayed-init.h [hunk 1]
```diff
@@ -7,23 +7,26 @@
...
+    constexpr struct EmptyDelayedInit
+    {
+    } delayed_init_empty;
+
     // implements the equivalent of function static initialization for an object
     template<class T>
     struct DelayedInit
     {
+        DelayedInit() = default;
+        DelayedInit(EmptyDelayedInit) { }
```
Comment (BillyONeal) [inline include/vcpkg/base/delayed-init.h:19]: These ctors have the same effect; not sure at what you're getting at here.

File: include/vcpkg/base/cache.h [hunk 2]
```diff
@@ -24,16 +27,28 @@ namespace vcpkg
         using is_callable = is_callable_impl<void, Callable, Args...>;
     }
 
+    // Thread-safe Cache
     template<class Key, class Value, class Compare = std::less<>>
     struct Cache
     {
+        // - It is safe to access entries from multiple threads
+        // - It is safe to access independent keys from the same thread (nested)
+        // - It is unsafe to recursively access the same key while initializing that key
```
Comment (BillyONeal) [inline include/vcpkg/base/cache.h:36]: This should be detectable/assertable and failfast?

**Comment by BillyONeal** [comment]
I'm not sure this is a good idea: even if we make everything in vcpkg.exe thread safe, the tools cache location on disk might not like this partying happening? Is there a reason we can't figure out what the tools are in advance before doing the pipelined part after parsing the cache configuration strings?

**Comment by autoantwort** [comment]
> the tools cache location on disk might not like this partying happening

What do you mean by that?

**Comment by autoantwort** [comment]
We could use file locking (https://cmake.org/cmake/help/latest/command/file.html#locking) to synchronize or use different write dirs.

**Comment by ras0219-msft** [comment]
> We could use file locking ...

And since we're written in the same language as CMake itself we can be 100% sure we're using the same locking code ;). Though I think before going further with this approach, we would really need to migrate all of the interactions with `downloads/` into the vcpkg tool via x-download etc.

>Is there a reason we can't figure out what the tools are in advance before doing the pipelined part after parsing the cache configuration strings?

I think this is probably the right way to go. It would be incredibly straightforward to add a `virtual void IBinaryPushProvider::fetch_tools(const ToolCache&);` which gets called once on the first `push_success()` call.

File: include/vcpkg/base/delayed-init.h (hunk 1 above)
Comment (ras0219-msft) [reply include/vcpkg/base/delayed-init.h:19]: `delayed_init_empty` is a safe, copiable placeholder that can be passed through `.emplace()` functions.

**Comment by BillyONeal** [comment]
> We could use file locking (https://cmake.org/cmake/help/latest/command/file.html#locking) to synchronize or use different write dirs.

Yes, but… why? Can’t we just do a pass over the configured providers just after parsing and ask them to fetch any tools or similar they need?

**Comment by autoantwort** [comment]
> I think this is probably the right way to go. It would be incredibly straightforward to add a virtual void IBinaryPushProvider::fetch_tools(const ToolCache&); which gets called once on the first push_success() call.

+1 
Or on the first `prefetch` or `precheck` call

**Review by Usrwrr — APPROVED** [review]
Comment

**Pull request closed without merging**